qiskit_ibm_runtime/execution_span.py

```diff
@@ -0,0 +1,94 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Execution span classes."""
+
+from typing import Union, Iterable, Sequence, TypeVar
+from datetime import datetime
+from dataclasses import dataclass
+
+
+"""The format accepted by ``numpy.ndarray.__getitem__()``."""
+SliceType = tuple[Union[slice, int, list[int]], ...]
+ExecutionSpanT = TypeVar("ExecutionSpanT", bound="ExecutionSpan")
+
+
+@dataclass(frozen=True, slots=True)
+class ExecutionSpan:
+    """Stores an execution time span for a subset of job data."""
+
+    start: datetime
+    """The start time of the span, in UTC."""
+
+    stop: datetime
+    """The stop time of the span, in UTC."""
+
+    data_slices: dict[int, SliceType]
+    r"""Which data have dependence on this execution span.
+
+    Data from the primitives are array-based, with every field in a 
+    :class:`~PubResult`\s :class:`~.DataBin` sharing the same base shape.
+    Therefore, the format of this field is a mapping from pub indexes to 
+    the same format accepted by 
+    NumPy slicing, where each value indicates which slice of each field in the
+    data bin depend on raw data collected during this execution span.
+    """
+
+    @property
+    def duration(self) -> float:
+        return (self.stop - self.start).seconds
+
+    def contains_pub(self, pub_idx: int | Iterable[int]):
+        pub_idx = {pub_idx} if isinstance(pub_idx, int) else set(pub_idx)
+        return not pub_idx.isdisjoint(self.data_slices)
+
+    def filter_by_pub(self, pub_idx: int | Iterable[int]) -> ExecutionSpanT:
+        pub_idx = {pub_idx} if isinstance(pub_idx, int) else set(pub_idx)
+        slices = {idx: sl for idx, sl in self.data_slices.items() if idx in pub_idx}
+        return ExecutionSpan(self.start, self.stop, slices)
+
+
+class ExecutionSpanCollection(Sequence[ExecutionSpan]):
+    """A collection of timings for the PUB result."""
+
+    def __init__(self, spans: Iterable[ExecutionSpan]):
+        self._spans = spans
```

Comment:
I think we can either inherit from `Sequence[ExecutionSpan]` or store `_spans`. I prefer the second option, @ihincks what do you think? Anyway we don't need both.

Reply:
Let's not inherit from `Sequence[ExecutionSpan]`. (Depending on whether you get it from abc.collections or typing, it would buy you different things)

Let's just implement `__iter__ -> Iterable[ExecutionSpan]` , `__len__ -> int`, and

```
@typing.overload
def __getitem__(self, idxs: slice | list[int]) -> ExecutionCollection: ...

@typing.overload
def __getitem__(self, idxs: int) -> ExecutionSpan: ...
```

which will make it implicitly obey the sequence protocol.